lib/Optimizer/Transforms/StateDecomposer.h

```diff
@@ -0,0 +1,175 @@
+#include "mlir/Transforms/RegionUtils.h"
+#include <span>
+
+namespace cudaq::details {
+
+/// @brief Converts angles of a uniformly controlled rotation to angles of
+/// non-controlled rotations.
+std::vector<double> convertAngles(const std::span<double> alphas);
+
+/// @brief Return the control indices dictated by the gray code implementation.
+///
+/// Here, numBits is the number of controls.
+std::vector<std::size_t> getControlIndices(std::size_t numBits);
+
+/// @brief Return angles required to implement a uniformly controlled z-rotation
+/// on the `kth` qubit.
+std::vector<double> getAlphaZ(const std::span<double> data,
+                              std::size_t numQubits, std::size_t k);
+
+/// @brief Return angles required to implement a uniformly controlled y-rotation
+/// on the `kth` qubit.
+std::vector<double> getAlphaY(const std::span<double> data,
+                              std::size_t numQubits, std::size_t k);
+} // namespace cudaq::details
+
+class StateGateBuilder {
+public:
+  StateGateBuilder(mlir::OpBuilder &b, mlir::Location &l, mlir::Value &q)
+      : builder(b), loc(l), qubits(q) {}
+
+  template <typename Op>
+  void applyRotationOp(double theta, std::size_t target) {
+    auto qubit = createQubitRef(target);
+    auto thetaValue = createAngleValue(theta);
+    builder.create<Op>(loc, thetaValue, mlir::ValueRange{}, qubit);
+  };
+
+  void applyX(std::size_t control, std::size_t target) {
+    auto qubitC = createQubitRef(control);
+    auto qubitT = createQubitRef(target);
+    builder.create<quake::XOp>(loc, qubitC, qubitT);
+  };
+
+private:
+  mlir::Value createQubitRef(std::size_t index) {
+    if (qubitRefs.contains(index)) {
+      return qubitRefs[index];
+    }
+
+    auto indexValue = builder.create<mlir::arith::ConstantIntOp>(
+        loc, index, builder.getIntegerType(64));
+    auto ref = builder.create<quake::ExtractRefOp>(loc, qubits, indexValue);
+    qubitRefs[index] = ref;
+    return ref;
+  }
+
+  mlir::Value createAngleValue(double angle) {
+    return builder.create<mlir::arith::ConstantFloatOp>(
+        loc, llvm::APFloat{angle}, builder.getF64Type());
+  }
+
+  mlir::OpBuilder &builder;
+  mlir::Location &loc;
+  mlir::Value &qubits;
+
+  std::unordered_map<std::size_t, mlir::Value> qubitRefs =
+      std::unordered_map<std::size_t, mlir::Value>();
+};
+
+class StateDecomposer {
+public:
+  StateDecomposer(StateGateBuilder &b, std::span<std::complex<double>> a)
+      : builder(b), amplitudes(a), numQubits(log2(a.size())) {}
+
+  /// @brief Decompose the input state vector data to a set of controlled
+  /// operations and rotations. This function takes as input a `OpBuilder`
+  /// and appends the operations of the decomposition to its internal
+  /// representation. This implementation follows the algorithm defined in
+  /// `https://arxiv.org/pdf/quant-ph/0407010.pdf`.
+  void decompose() {
+
+    // Decompose the state into phases and magnitudes.
```

Comment:
Is it possible to move any more of this code out of the header file?

On the third hand, all of this could be in the StatePreparation file.

Reply:
moved to StatePreparation.cpp